### DISCUSSION TITLE
Exclude env vars from bundle

### DISCUSSION THREAD

**Discussion by hobadams** [vercel/next.js#51212] [Help]
### Summary

Is it possible to build a Next Js app (v13.4) where all env vars are excluded from the bundle?

If a component is a server component will the var be grabbed from the server or will it be bundled in the build if possible?

I’m not talking about NEXT_PUBLIC vars.

**Comment by devjiwonchoi** [comment]
Just to make sure, could you clarify `exclude env vars from the bundle`?
One sure thing is that the `.env` vars run on a Node environment therefore the Server Components won't be affected since the vars will be grabbed from the server.

> This loads process.env.DB_HOST, process.env.DB_USER, and process.env.DB_PASS into the Node.js environment automatically allowing you to use them in Route Handlers.

## Reference
- [Environment Variables](https://nextjs.org/docs/app/building-your-application/configuring/environment-variables)

**Comment by icyJoseph** [comment]
This is tricky to explain and get a good grip on... 

Server Components have NO client side JS bundle, so that should never arrive to the browser. But will it be inlined?

This code:

```js
const foo = async () => {
  try {
    await fetch("-bad-url", { cache: "no-store" });
  } catch (_) {
    return process.env.NEXT_PUBLIC_TOKEN;
  }
};

const baz = async () => {
  try {
    await fetch("-bad-url-", { cache: "no-store" });
  } catch (_) {
    return process.env.PRIVATE_TOKEN;
  }
};

export default async function Home() {
  const x = await foo();
  const y = await baz();
  return (
    <div>
      <p>public: {x}</p>
      <p>private: {y}</p>
    </div>
  );
}

```

Becomes

```js
const foo = async ()=>{
    try {
        await fetch("-bad-url", {
            cache: "no-store"
        });
    } catch (_) {
        return "__foo__";
    }
};
const baz = async ()=>{
    try {
        await fetch("-bad-url-", {
            cache: "no-store"
        });
    } catch (_) {
        return process.env.PRIVATE_TOKEN;
    }
};
async function Home() {
    const x = await foo();
    const y = await baz();
    return /*#__PURE__*/ (0,react_jsx_runtime__WEBPACK_IMPORTED_MODULE_0__.jsxs)("div", {
        children: [
            /*#__PURE__*/ (0,react_jsx_runtime__WEBPACK_IMPORTED_MODULE_0__.jsxs)("p", {
                children: [
                    "public: ",
                    x
                ]
            }),
```

So the Private Token is not inlined into the JavaScript, and this page is labeled, `λ`, which means it is not pre-rendered, it'll be rendered on demand (no HTML is made for it at build time).

Let' s step into the murky waters, without `foo` and `baz` functions, just raw inline of the `env vars`:

```jsx
export default async function Home() {
  return (
    <div>
      <p>public: {process.env.NEXT_PUBLIC_TOKEN}</p>
      <p>private: {process.env.PRIVATE_TOKEN}</p>
    </div>
  );
}
```

Outputs, this JavaScript, as expected:

```js
async function Home() {
    return /*#__PURE__*/ (0,react_jsx_runtime__WEBPACK_IMPORTED_MODULE_0__.jsxs)("div", {
        children: [
            /*#__PURE__*/ (0,react_jsx_runtime__WEBPACK_IMPORTED_MODULE_0__.jsxs)("p", {
                children: [
                    "public: ",
                    "__foo__"
                ]
            }),
            /*#__PURE__*/ (0,react_jsx_runtime__WEBPACK_IMPORTED_MODULE_0__.jsxs)("p", {
                children: [
                    "private: ",
                    process.env.PRIVATE_TOKEN
                ]
            })
        ]
    });
}
```

But! It also outputs this pre-rendered HTML:

```html
<!DOCTYPE html>
<html lang="en">
  <head>
     <!-- Removed head contents for visibility -->
  </head>
  <body class="__className_0ec1f4">
    <div>
      <p>public:
        <!-- -->__foo__
      </p>
      <p>private:
        <!-- -->__bar__
      </p>
    </div>
  <!-- Removed scripts for visibility -->
  </body>
</html>
```

Since `Home` in this case, is a Server Component, the `__bar__` bit is never hydrated, so it never becomes undefined or anything like that. A leak has been made. Clients without JS would see it regardless, so it is not a good idea at all to inline env vars into JSX. 

> This also happened in Pages Directory, as far back as next@11 IIRC. But the hydration process removed the value, though it was there to being with, which is a problem

Though why would anyone inline a secret into HTML. It is also a bit of a forced scenario from my side.

Going back to `foo` and `baz`. Using no-store opts them into dynamic rendering, so there is no pre-generated HTML, and hence no leak.

These **also inline** the PRIVATE_TOKEN:

```js

export default async function Home() {
  // const x = await foo();
  const y = process.env.PRIVATE_TOKEN;
  return (
    <div>
      {/* <p>public: {x}</p> */}
      <p>private: {y}</p>
    </div>
  );
}
```

```js

const getPT = () => process.env.PRIVATE_TOKEN;

export default async function Home() {
  const y = getPT();
  return (
    <div>
      <p>private: {y}</p>
    </div>
  );
}

```

So it seems like our only hope is to wrap them in fetch, and use either `no-store`, `{next: {revalidate: 0}}` etc... What we want to do, is prevent the Server Component pre-rendering optimisation, that is, the page needs `λ` over `○`.

I've found that for white-label apps, or apps that want to do the 12-steps thingy, this is kind of annoying... 😓 My approach, from way back, has been to promisify reading variables, to prevent them from easily ending up on JSX and so inlined into the HTML, however, now you can await within a Server Component... so fetch caching comes to the rescue.

In Pages dir, if you had pre-rendered in some way, you'd leak too...

```jsx
export default function Home() {
  return (
    <div>
      Home
      <div>{process.env.PRIVATE_TOKEN}</div>
    </div>
  );
}
```

Pre-renders:

```html
<!DOCTYPE html>
<html lang="en">
  <head>
   
  </head>
  <body>
    <div id="__next">
      <div>Home <div>__bar__</div>
      </div>
    </div>
    <script id="__NEXT_DATA__" type="application/json">
      {
        "props": {
          "pageProps": {}
        },
        "page": "/",
        "query": {},
        "buildId": "yctDBoBDtthQIRbqvgL4n",
        "nextExport": true,
        "autoExport": true,
        "isFallback": false,
        "scriptLoader": []
      }
    </script>
  </body>
</html>
```

And that's why I used an async getter and never access process.env directly. `PRIVATE_TOKEN` won't be in the JavaScript for this Pages Dir page though, so if you run the sever, you "won't see it", because the hydration process drops it.

**Comment by hobadams** [comment]
Thank you @icyJoseph and @devjiwonchoi for your answers.

We've currently solved this by excluding some build files (in our docker ignore). This isn't a future proof solution as we need to know exactly the components that need files excluding.  luckily for us it's only 1 component giving us grief.

We have a logout component and our docker ignore looks like this

```
**/logout.body
**/logout.meta
```

**Comment by aliev** [comment]
> ### Summary
> Is it possible to build a Next Js app (v13.4) where all env vars are excluded from the bundle?
> 
> If a component is a server component will the var be grabbed from the server or will it be bundled in the build if possible?
> 
> I’m not talking about NEXT_PUBLIC vars.
> 
> ### Additional information
> _No response_
> 
> ### Example
> _No response_

I'm new to Next.js, but maybe the solution I came up with can help you too.

I was concerned about my environment variables potentially ending up in the build (whether a Docker image or a JavaScript bundle), so I created this workaround for myself (ignore the asynchronous function part – in my case, it works even without `async`):

```typescript
/**
 * Retrieves a secret environment variable.
 */
export async function getSecretEnv(
  envName: string,
): Promise<string> {
  if (process.env.BUILD_MODE !== undefined) {
    return process.env[envName] === undefined ? "" : process.env[envName];
  }

  if (process.env[envName] === undefined) {
    throw new Error(`${envName} environment variable is not set`);
  }

  return process.env[envName];
}
```

To make this work, we need to update `package.json`:

```json
...
"build": "BUILD_MODE=1 next build",
...
```

The main idea is as follows: during compilation (or build), we ignore any missing environment variables by simply returning an empty string. To recognize the build stage, we add an extra environment variable: `BUILD_MODE=1` that we pass in during the build.

At runtime (the second condition), we just throw an error if an environment variable is missing.

When might this approach be useful?

1. You want to pass environment variables dynamically when launching your image or application.
2. You’re concerned that environment variables (secrets) might end up in the build.

**Comment by olamide226** [comment]
One pattern that helps here is to stop treating all env vars as one bucket.

The risky part is when a value is used in code that can be evaluated during build/pre-render, or when a supposedly private value accidentally crosses into JSX/HTML/client JS. That makes it easy to leak something by mistake.

REP tries to make that boundary explicit:

https://github.com/RuachTech/rep

It uses three prefixes:

```bash
REP_PUBLIC_API_URL=https://api.example.com          # browser-visible, plaintext
REP_SENSITIVE_ANALYTICS_KEY=...                     # browser-delivered, encrypted
REP_SERVER_DB_PASSWORD=...                          # never sent to browser
```

So instead of using `process.env` directly from client-facing code, browser config is read from a runtime payload:

```ts
import { rep } from '@rep-protocol/sdk'

const apiUrl = rep.get('API_URL')
```

The gateway injects the browser payload as inert JSON:

```html
<script id="__rep__" type="application/json">...</script>
```

The useful security property is that `REP_SERVER_*` values are not included in that payload at all. They stay in the gateway/server process and never reach the browser document.

REP also has startup guardrails for accidental leaks. For example, if someone does this:

```bash
REP_PUBLIC_STRIPE_SECRET_KEY=sk_live_xxx
REP_PUBLIC_AWS_KEY=AKIA...
REP_PUBLIC_TOKEN=eyJhbGciOi...
```

those values are flagged because public values are scanned for common secret formats and high-entropy strings. In strict mode, the gateway can fail startup instead of silently shipping the value to the browser.

For a Docker/standalone setup, this keeps the image portable without baking browser config into the Next build:

```sh
NODE_ENV=production node /app/server.js &

exec rep-gateway \
  --mode proxy \
  --port 8080 \
  --upstream localhost:3000
```

Then runtime config is passed when the container starts:

```bash
docker run --rm -p 8080:8080 \
  -e REP_PUBLIC_API_URL=https://api.example.com \
  -e REP_SERVER_DB_PASSWORD='server-only-secret' \
  -e REP_GATEWAY_STRICT=true \
  my-next-app:latest
```

Important distinction: true secrets should still be `REP_SERVER_*` or normal server-only env vars. `REP_SENSITIVE_*` is for values that intentionally need to be delivered to the browser but should not sit as plaintext in page source. It is not a reason to send database passwords or private signing keys to client code.

So REP does not change how Next.js optimizes Server Components or pre-renders pages. It gives deployments a separate runtime config layer with explicit public/sensitive/server classification, plus secret scanning to catch accidental public leaks before the app starts.
